parallel.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2023 Stock Parfait
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"sync"
+)
+
+type contextKey int
+
+const (
+	parallelKey contextKey = iota
+)
+
+// TestSerialize forces the number of workers in Map to be 1, thereby running
+// jobs serially and in the strict FIFO order. This helps make tests
+// deterministic.
+func TestSerialize(ctx context.Context) context.Context {
+	return context.WithValue(ctx, parallelKey, true)
+}
+
+func isSerialized(ctx context.Context) bool {
+	v, ok := ctx.Value(parallelKey).(bool)
+	return ok && v
+}
+
+// Job is a unit of work to be done, which returns a result.
+type Job[T any] func() T
+
+type mapIter[T any] struct {
```

Comment:
Rename to parallelMapIter, to disambiguate with a sequential `Map` implementation.